src/components/ResourceList/ResourceList.tsx

```diff
@@ -791,29 +509,202 @@ class ResourceListInner extends React.Component<CombinedProps, State> {
-  });
-}
+  const bulkActionsMarkup = isSelectable ? (
+    <div className={styles.BulkActionsWrapper}>
+      <BulkActions
+        label={bulkActionsLabel()}
+        accessibilityLabel={bulkActionsAccessibilityLabel()}
+        selected={bulkSelectState()}
+        onToggleAll={handleToggleAll}
+        selectMode={selectMode}
+        onSelectModeToggle={handleSelectMode}
+        promotedActions={promotedBulkActions}
+        paginatedSelectAllAction={paginatedSelectAllAction()}
+        paginatedSelectAllText={paginatedSelectAllText()}
+        actions={bulkActions}
+        disabled={loading}
+        smallScreen={smallScreen}
+      />
+    </div>
+  ) : null;
+
+  const filterControlMarkup = filterControl ? (
+    <div className={styles.FiltersWrapper}>{filterControl}</div>
+  ) : null;
+
+  const sortingSelectMarkup =
+    sortOptions && sortOptions.length > 0 && !alternateTool ? (
+      <div className={styles.SortWrapper}>
+        <Select
+          label={i18n.translate('Polaris.ResourceList.sortingLabel')}
+          labelInline={!smallScreen}
+          labelHidden={smallScreen}
+          options={sortOptions}
+          onChange={onSortChange}
+          value={sortValue}
+          disabled={selectMode}
+        />
+      </div>
+    ) : null;
 
-function defaultIdForItem(item: any, index: number) {
-  return Object.prototype.hasOwnProperty.call(item, 'id')
-    ? item.id
-    : index.toString();
-}
+  const alternateToolMarkup =
+    alternateTool && !sortingSelectMarkup ? (
+      <div className={styles.AlternateToolWrapper}>{alternateTool}</div>
+    ) : null;
 
-function isSmallScreen() {
-  return typeof window === 'undefined'
-    ? false
-    : window.innerWidth < SMALL_SCREEN_WIDTH;
-}
+  const headerTitleMarkup = (
+    <div className={styles.HeaderTitleWrapper} testID="headerTitleWrapper">
+      {headerTitle()}
+    </div>
+  );
+
+  const selectButtonMarkup = isSelectable ? (
+    <div className={styles.SelectButtonWrapper}>
+      <Button
+        disabled={selectMode}
+        icon={EnableSelectionMinor}
+        onClick={() => handleSelectMode(true)}
+      >
+        {i18n.translate('Polaris.ResourceList.selectButtonText')}
+      </Button>
+    </div>
+  ) : null;
+
+  const checkableButtonMarkup = isSelectable ? (
+    <div className={styles.CheckableButtonWrapper}>
+      <CheckableButton
+        accessibilityLabel={bulkActionsAccessibilityLabel()}
+        label={headerTitle()}
+        onToggleAll={handleToggleAll}
+        plain
+        disabled={loading}
+      />
+    </div>
+  ) : null;
+
+  const needsHeader =
+    isSelectable || (sortOptions && sortOptions.length > 0) || alternateTool;
+
+  const headerWrapperOverlay = loading ? (
+    <div className={styles['HeaderWrapper-overlay']} />
+  ) : null;
+
+  const showEmptyState = filterControl && !itemsExist && !loading;
+
+  const headerMarkup = !showEmptyState &&
+    (showHeader || needsHeader) &&
+    listRef.current && (
```

Comment:
Really thought we could check the items here. Turns out it doesn't work and we need to check listRef.current for the header to stick.

Reply:
Huh, really? It worked when I was testing locally